File: packages/upay-php/src/Resources/Transactions.php
<?php

namespace Upay\Resources;

use Upay\HttpClient;

class Transactions
{
    private HttpClient $http;
    
    public function __construct(HttpClient $http)
    {
        $this->http = $http;
    }
    
    public function create(array $data): array
    {
        // Validação básica
        if (empty($data['product']) || strlen(trim($data['product'])) === 0) {
            throw new \InvalidArgumentException('Produto é obrigatório');
        }
        
        if (empty($data['amountCents']) || $data['amountCents'] < 100) {
            throw new \InvalidArgumentException('Valor mínimo é R$ 1,00 (100 centavos)');
        }
        
        if (!empty($data['client']) && empty($data['client']['email'])) {
            throw new \InvalidArgumentException('Email do cliente é obrigatório');
        }
        
        return $this->http->post('/transactions', $data);
    }
    
    public function list(?array $params = null): array
    {
        $response = $this->http->get('/transactions', $params);
        
        return [
            'data' => $response['transactions'] ?? $response['data'] ?? [],
            'pagination' => $response['pagination'] ?? ['total' => 0, 'page' => 1, 'limit' => 10]
        ];
    }
    
    public function get(string $id): array
    {
        if (empty($id)) {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        return $this->http->get("/transactions/{$id}");
    }
    
    public function process(string $id, ?array $paymentData = null): array
    {
        if (empty($id)) {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        return $this->http->post("/transactions/{$id}/process", $paymentData);
    }
    
    public function capture(string $id): array
    {
        if (empty($id)) {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        return $this->http->post("/transactions/{$id}/capture");
    }
    
    public function cancel(string $id): array
    {
        if (empty($id)) {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        return $this->http->post("/transactions/{$id}/cancel");
    }
    
    public function refund(string $id, ?int $amountCents = null): array
    {
        if (empty($id)) {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        $data = [];
        if ($amountCents !== null) {
            $data['amountCents'] = $amountCents;
        }
        
        return $this->http->post("/transactions/{$id}/refund", $data);
    }
}

File: packages/upay-python/upay/resources/transactions.py
"""
Recurso de Transações
"""

from typing import Optional, Dict, Any
from ..http import HttpClient


class TransactionsResource:
    """Recurso para gerenciar Transações"""
    
    def __init__(self, http: HttpClient):
        self.http = http
    
    def create(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Cria uma nova transação
        
        Args:
            data: Dados da transação
                - product: Nome do produto (obrigatório)
                - amountCents: Valor em centavos (obrigatório, min 100)
                - paymentMethod: Método de pagamento (PIX, CREDIT_CARD, BOLETO)
                - clientId: ID do cliente
                - client: Dados do cliente (se não tiver clientId)
                - paymentLinkId: ID do link de pagamento
                - metadata: Metadados adicionais
                - couponCode: Código do cupom
                
        Returns:
            Transação criada
        """
        # Validação básica
        if not data.get("product") or len(str(data["product"]).strip()) == 0:
            raise ValueError("Produto é obrigatório")
        
        if not data.get("amountCents") or data.get("amountCents", 0) < 100:
            raise ValueError("Valor mínimo é R$ 1,00 (100 centavos)")
        
        if data.get("client") and not data["client"].get("email"):
            raise ValueError("Email do cliente é obrigatório")
        
        return self.http.post("/transactions", data)
    
    def list(
        self,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        order_by: Optional[str] = None,
        order_direction: Optional[str] = None,
        status: Optional[str] = None,
        payment_method: Optional[str] = None,
        client_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Lista transações
        
        Args:
            page: Número da página
            limit: Limite de itens por página
            cursor: Cursor para paginação
            order_by: Campo para ordenação
            order_direction: Direção da ordenação (ASC ou DESC)
            status: Filtrar por status
            payment_method: Filtrar por método de pagamento
            client_id: Filtrar por cliente
            
        Returns:
            Dicionário com 'data' (lista) e 'pagination'
        """
        params = {
            "page": page,
            "limit": limit,
            "cursor": cursor,
            "orderBy": order_by,
            "orderDirection": order_direction,
            "status": status,
            "method": payment_method,
            "clientId": client_id,
        }
        
        response = self.http.get("/transactions", params)
        
        # Mapear resposta: { message, transactions, pagination } -> { data, pagination }
        return {
            "data": response.get("transactions") or response.get("data") or [],
            "pagination": response.get("pagination") or {"total": 0, "page": 1, "limit": 10}
        }
    
    def get(self, transaction_id: str) -> Dict[str, Any]:
        """
        Obtém uma transação por ID
        
        Args:
            transaction_id: ID da transação
            
        Returns:
            Transação
        """
        if not transaction_id:
            raise ValueError("ID é obrigatório")
        
        return self.http.get(f"/transactions/{transaction_id}")
    
    def process(
        self,
        transaction_id: str,
        payment_data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Processa o pagamento de uma transação
        
        Args:
            transaction_id: ID da transação
            payment_data: Dados do pagamento (cardData, installments)
            
        Returns:
            Transação processada
        """
        if not transaction_id:
            raise ValueError("ID é obrigatório")
        
        return self.http.post(f"/transactions/{transaction_id}/process", payment_data)
    
    def capture(self, transaction_id: str) -> Dict[str, Any]:
        """
        Captura uma transação autorizada (Pagar.me)
        
        Args:
            transaction_id: ID da transação
            
        Returns:
            Transação capturada
        """
        if not transaction_id:
            raise ValueError("ID é obrigatório")
        
        return self.http.post(f"/transactions/{transaction_id}/capture")
    
    def cancel(self, transaction_id: str) -> Dict[str, Any]:
        """
        Cancela uma transação pendente
        
        Args:
            transaction_id: ID da transação
            
        Returns:
            Transação cancelada
        """
        if not transaction_id:
            raise ValueError("ID é obrigatório")
        
        return self.http.post(f"/transactions/{transaction_id}/cancel")
    
    def refund(
        self,
        transaction_id: str,
        amount_cents: Optional[int] = None
    ) -> Dict[str, Any]:
        """
        Estorna uma transação paga
        
        Args:
            transaction_id: ID da transação
            amount_cents: Valor a estornar em centavos (opcional, estorna tudo se não informado)
            
        Returns:
            Transação estornada
        """
        if not transaction_id:
            raise ValueError("ID é obrigatório")
        
        data = {}
        if amount_cents is not None:
            data["amountCents"] = amount_cents
        
        return self.http.post(f"/transactions/{transaction_id}/refund", data)
